src/middlewared/middlewared/plugins/pool.py

```diff
@@ -2749,9 +2749,11 @@ def query(self, filters=None, options=None):
                 zfsfilters.append(['id', '=', f[2]])
                 break
 
+        extra = options.get('extra', {})
+        extra.setdefault('flat', True)
         return filter_list(
             self.__transform(self.middleware.call_sync(
-                'zfs.dataset.query', zfsfilters, {'extra': {'flat': options.get('extra', {}).get('flat', True)}})
+                'zfs.dataset.query', zfsfilters, {'extra': extra})
```

Comment:
Let's please not pass the complete `extra` payload to `zfs.dataset.query` as there are configuration knobs there which can potentially cause further bugs/issues as they can alter the output received from the internal service. I suppose you want `recursive` to be configurable, in this case let's please just add `recursive` to the extra argument like we were doing for `flat`